x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/TransportVersionClusterStateUpgradeIT.java

```diff
@@ -12,70 +12,135 @@
+            client().performRequest(new Request("GET", "/_cluster/state" + randomFrom("", "/nodes") + randomFrom("", "?local")))
         );
-        Map<?, TransportVersion> tvs = ((List<?>) localResponse.get("transport_versions")).stream()
-            .map(o -> (Map<?, ?>) o)
-            .collect(Collectors.toMap(m -> m.get("node_id"), m -> TransportVersion.fromString(m.get("transport_version").toString())));
+        final var description = clusterState.toString();
+
+        final var nodeIds = clusterState.evaluateMapKeys("nodes");
+        final Map<String, Version> versionsByNodeId = Maps.newHashMapWithExpectedSize(nodeIds.size());
+        for (final var nodeId : nodeIds) {
+            versionsByNodeId.put(nodeId, Version.fromString(clusterState.evaluate("nodes." + nodeId + ".version")));
+        }
+
+        final var hasTransportVersions = clusterState.evaluate("transport_versions") != null;
+        final var hasNodesVersions = clusterState.evaluate("nodes_versions") != null;
+        assertFalse(description, hasNodesVersions && hasTransportVersions);
 
-        for (var ver : vs.entrySet()) {
-            if (ver.getValue().after(Version.V_8_8_0)) {
-                assertThat(
-                    "Node " + ver.getKey() + " should have an inferred transport version",
-                    tvs.get(ver.getKey()),
-                    equalTo(TransportVersions.V_8_8_0)
-                );
+        switch (CLUSTER_TYPE) {
+            case OLD -> {
+                if (UPGRADE_FROM_VERSION.before(VERSION_INTRODUCING_TRANSPORT_VERSIONS)) {
+                    assertFalse(description, hasTransportVersions);
+                    assertFalse(description, hasNodesVersions);
+                } else if (UPGRADE_FROM_VERSION.before(VERSION_INTRODUCING_NODES_VERSIONS)) {
+                    assertTrue(description, hasTransportVersions);
+                    assertFalse(description, hasNodesVersions);
+                } else {
+                    assertFalse(description, hasTransportVersions);
+                    assertTrue(description, hasNodesVersions);
+                }
+            }
+            case MIXED -> {
+                if (UPGRADE_FROM_VERSION.before(VERSION_INTRODUCING_TRANSPORT_VERSIONS)) {
+                    assertFalse(description, hasTransportVersions);
+                } else if (UPGRADE_FROM_VERSION.before(VERSION_INTRODUCING_NODES_VERSIONS)) {
+                    assertTrue(description, hasNodesVersions || hasTransportVersions);
+                } else {
+                    assertFalse(description, hasTransportVersions);
+                    assertTrue(description, hasNodesVersions);
+                }
+            }
+            case UPGRADED -> {
+                assertFalse(description, hasTransportVersions);
+                assertTrue(description, hasNodesVersions);
+                assertTrue(description, versionsByNodeId.values().stream().allMatch(v -> v.equals(Version.CURRENT)));
```

Comment:
`everyItem` matcher is a more declarative way of doing this

Reply:
++ TIL